cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,353 @@
+		"The name of the kubeconfig context to use")
+	cmd.PersistentFlags().BoolVarP(&args.yes, "yes", "y", false,
+		"If yes, skips the prompting confirmation for value changes in this upgrade")
+	cmd.PersistentFlags().BoolVarP(&args.wait, "wait", "w", false,
+		"Wait, if set will wait until all Pods, Services, and minimum number of Pods "+
+			"of a Deployment are in a ready state before the command exits. "+
+			"It will wait for a maximum duration of "+strconv.Itoa(upgradeWaitSecWhenApply)+" seconds")
+	cmd.PersistentFlags().BoolVar(&args.force, "force", false,
+		"Apply the upgrade without eligibility checks and testing for changes "+
+			"in profile default values")
+	cmd.PersistentFlags().StringVarP(&args.versionsURI, "versionsURI", "u",
+		versionsMapURL, "URI for operator versions to Istio versions map")
+}
+
+// Upgrade command upgrades Istio control plane in-place with eligibility checks
+func Upgrade() *cobra.Command {
+	macArgs := &upgradeArgs{}
+	rootArgs := &rootArgs{}
+	cmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "Upgrade Istio control plane in-place",
+		Long: "The mesh upgrade command checks for upgrade version eligibility and," +
+			" if eligible, upgrades the Istio control plane components in-place. Warning: " +
+			"traffic may be disrupted during upgrade. Please ensure PodDisruptionBudgets " +
+			"are defined to maintain service continuity.",
+		Example: `mesh upgrade`,
+		RunE: func(cmd *cobra.Command, args []string) (e error) {
+			l = newLogger(rootArgs.logToStdErr, cmd.OutOrStdout(), cmd.OutOrStderr())
+			return upgrade(rootArgs, macArgs)
+		},
+	}
+	addFlags(cmd, rootArgs)
+	addUpgradeFlags(cmd, macArgs)
+	return cmd
+}
+
+// upgrade is the main function for Upgrade command
+func upgrade(rootArgs *rootArgs, args *upgradeArgs) (err error) {
+	initLogsOrExit(rootArgs)
+	l.logAndPrintf("Client - istioctl version: %s\n", opversion.OperatorVersionString)
+
+	targetValues := genValuesFromFile(args.inFilename, args.force)
+	targetICPS := genICPSFromFile(args.inFilename, args.force)
+	targetVersion := targetICPS.GetTag()
+	l.logAndPrintf("Upgrade - target version: %s\n", targetVersion)
+
+	kubeClient := getKubeExecClient(args.kubeConfigPath, args.context)
+	//TODO(elfinhe): support components distributed in multiple namespaces
+	istioNamespace := targetICPS.GetDefaultNamespace()
+	currentVer := retrieveControlPlaneVersion(kubeClient, istioNamespace)
+	currentValues := readValuesFromInjectorConfigMap(kubeClient, istioNamespace)
+
+	if !args.force {
+		checkSupportedVersions(currentVer, targetVersion, args.versionsURI)
+		checkUpgradeValues(currentValues, targetValues, args.yes)
+	}
+
+	runPreUpgradeHooks(kubeClient, istioNamespace,
+		currentVer, targetVersion, currentValues, targetValues, rootArgs.dryRun)
+	applyUpgradeManifest(args.inFilename, args.kubeConfigPath,
+		args.context, rootArgs.dryRun, rootArgs.verbose)
+	runPostUpgradeHooks(kubeClient, istioNamespace,
```

Comment:
all the functions in here should be returning error. you want to return error as much as possible since the code will be reused for the controller upgrade.